---
title: "Working with DNA sequences and features in R with Bioconductor"
output: 
    ioslides_presentation:
        widescreen: true
        smaller: true
        css: style.css
---

## Bioconductor

Bioconductor is repository for R packages relating to high-throughput biology (shotgun sequencing, microarrays, etc).

* Started in 2001, has developed alongside technology and growth of data
* Core set of types promotes interoperation<br>(Not just tidy data frames! Types have evolved over time to meet new needs.)
* Twice yearly release schedule
* Many contributors, but fairly good coding and documentation standards

## Flavours of R

### Base R

Functions and types that are always available in R.

### CRAN (14,917 packages as at September 2019)

"Comprehensive R Archive Network" - main source of R packages

```{r eval=F}
install.packages("glmnet")
install.packages("tidyverse")
```

### Bioconductor (1,741 packages as at September 2019)

```{r eval=F}
# Setup
install.packages("BiocManager")

# Install a Bioconductor package
BiocManager::install("limma")

# Check for out-of-date packages
BiocManager::install()
```


## What Bioconductor is for

**Peaks, regions, motifs in genome** <span style="color: #00aaaa">&larr; focus for today</span>

eg for ChIP-seq, ATAC-seq, variant calling

* `Biostrings`, `GenomicRanges`, `GenomicFeatures`, ...
<br>

**Differential expression analysis from microarray or RNA-seq**

* `SummarizedExperiment`, `limma`, `edgeR`, `DESeq2`, ...
* Further packages to normalize, impute, batch correct, check quality, ...
<br>

**Visualization**

* `ComplexHeatmap`, `Gviz`, `ggbio`, ...
<br>

**Single-cell sequencing**

* `SingleCellExperiment`, ...
<br>

**Statistical methods for $p \gg n$ data**

...


## More types

Can do a lot in R with vectors and data frames.

Using Bioconductor will mean building familiarity with further types:

* `matrix`, `list`
<br>
<br>

* `DNAString`, `DNAStringSet`
* `GRanges`
<br>
<br>

* `Seqinfo`
* `TxDb`, `EnsDb`
<br>
<br>

* `OrgDb`, `VCF`, `SummarizedExperiment`, `DelayedArray`, ...

## S4 types

Most base R packages use "S3" types. Data frames, tibbles, and linear model fits are examples of these.

* essentially lists
* if necessary, peek inside with `$`

<br>
Bioconductor uses "S4" types, including it's own data frame (`DataFrame`) and list types (`SimpleList`, `GRangesList`, etc etc). If stuck, can almost always convert to base R types with `as.data.frame`, `as.list`, `as.character`, `as.numeric`.

* use accessor functions such as `seqnames`, `start`, `end`, `width`, `nchars`
* if absolutely necessary, peek inside with `@`

<br>
[Stuart Lee's guide to S4 for the perplexed](https://stuartlee.org/post/content/post/2019-07-09-s4-a-short-guide-for-perplexed/)


## Command-line bioinformatics software

Not all bioinformatics software is an R package!

R's role will often be to massage your data into the form needed for command line tools, or to examine a tool's output.

<br>

One way to install command-line software is using the Conda package manager:

* [Install miniconda](https://docs.conda.io/projects/conda/en/latest/user-guide/install/)
* Use the [bioconda](https://anaconda.org/bioconda) channel

```
# Example
conda install -c bioconda meme
```

* Can be used on your own computer, or on a server or cluster.
* Doesn't need admin rights.


## Reference genomes and annotations

Analysis usually occurs in the context of:

* a genome assembly: DNA sequences of chromosomes 
    * `DNAStringSet` or `BSgenome` or `TwoBitFile`
* gene annotations: strand, ranges of positions in chromosome
    * `TxDb` or `EnsDb`

For model organisms such as human and mouse, there is usually a genome assembly that is updated infrequently, and then several sets of gene annotations (Ensembl, RefSeq, UCSC, etc) which are updated much more frequently.

<img src="../figures/igv.png" width="1000">


##

(Do workshop)


## Using Bioconductor

1. Find some useful packages
1. Read the vignettes
1. Read the reference documentation for specific functions you need
1. If you run into a funny class, check its documentation, work out the accessor functions, and in a pinch poke around its internals with `@` or try `as.data.frame`.

```{r eval=F}
?"GRanges-class"

methods(class="GRanges")
```

<br>

* The [Bioconductor website](https://bioconductor.org) includes material from many and various tutorials/vignettes/workflows
* [Most downloaded Bioconctor packages](http://bioconductor.org/packages/stats/)
* [Mike Love's Bioconductor cheat sheet](https://github.com/mikelove/bioc-refcard/blob/master/README.Rmd)
* [plyranges](https://bioconductor.org/packages/release/bioc/html/plyranges.html) provides a "tidy" way of working with GRanges with many powerful features. The author is doing their PhD here at Monash. [A plyranges workshop](https://bioconductor.github.io/BiocWorkshops/fluent-genomic-data-analysis-with-plyranges.html)
* [Bioconductor's Stack-Overflow-style support site](https://support.bioconductor.org/)


